traits/observation/tests/test_named_trait_observer.py

```diff
@@ -434,6 +434,75 @@ def test_remove_notifiers_after_trait_added(self):
         foo.value += 1
         self.assertEqual(handler.call_count, 0)
 
+    def test_remove_trait_then_add_trait_again(self):
+        # Test a scenario where a trait exists when the observer is hooked,
+        # but then the trait is removed, and then added back again, the
+        # observer is gone, because the CTrait is gone and the trait_added
+        # event is not fired for something already defined on the class.
+
+        # given
+        # the trait exists, we can set optional to false.
+        graph = create_graph(
+            create_observer(name="value1", notify=True, optional=False),
+        )
+        handler = mock.Mock()
+        foo = ClassWithTwoValue()
+        call_add_or_remove_notifiers(
+            object=foo, graph=graph, handler=handler, remove=False)
```

Comment:
This is the low level way to say `foo.observe("value1", handler)`.
A lot of the tests here work with lower level objects because the high level user-facing API did not exist then. This is still very close to the high level API. I will keep the local consistency for now.